jina/executors/decorators.py

```diff
@@ -255,11 +274,79 @@ def arg_wrapper(*args, **kwargs):
         return _batching
 
 
-def _get_size(data: Union[Iterator[Any], List[Any], np.ndarray], axis: int = 0) -> int:
-    if isinstance(data, np.ndarray):
-        total_size = data.shape[axis]
-    elif hasattr(data, '__len__'):
-        total_size = len(data)
+def batching_multi_input(func: Callable[[Any], np.ndarray] = None, *,
+                         batch_size: Union[int, Callable] = None,
+                         num_batch: Optional[int] = None,
+                         split_over_axis: int = 0,
+                         merge_over_axis: int = 0,
+                         slice_on: int = 1,
+                         num_data: int) -> Any:
+    """Split the input of a function into small batches and call :func:`func` on each batch
+    , collect the merged result and return. This is useful when the input is too big to fit into memory
+
+    :param func: function to decorate
+    :param batch_size: size of each batch
+    :param num_batch: number of batches to take, the rest will be ignored
+    :param split_over_axis: split over which axis into batches
+    :param merge_over_axis: merge over which axis into a single result
+    :param slice_on: the location of the data. When using inside a class,
+            ``slice_on`` should take ``self`` into consideration.
+    :param num_data: the number of data inside the arguments (starting from ``slice_on``)
+    :return: the merged result as if run :func:`func` once on the input.
+
+    Example:
+        .. highlight:: python
+        .. code-block:: python
+
+            class MultiModalExecutor:
+
+                @batching_multi_input(batch_size = 64, num_data=2)
+                def encode(self, *batches, **kwargs):
+                    batch_modality0 = batches[0]
+                    embed0 = _encode_modality(batch_modality0)
+                    batch_modality1 = batches[1]
+                    embed1 = _encode_modality(batch_modality0)
+    """
+
+    def _batching(func):
+        @wraps(func)
+        def arg_wrapper(*args, **kwargs):
+            # priority: decorator > class_attribute
+            # by default data is in args[1:] (self needs to be taken into account)
+            data_args = [args[slice_on + i] for i in range(0, num_data)]
+
+            data = list(zip(*data_args))
```

Comment:
One of the reasons to do batching is, keeping the memory footprint low. explicit conversion to a `list` enforces whatever iterators are behind `data_args` to be converted to a list in memory.

I think batching should treat whatever input it gets as a stream and not instantiate the full stream directly. Or if it does, we should change the docstring, that the caller function should already care about only calling the to-be-batched function with in-memory data.

Reply:
That was one of my first concerns when doing it, but did not see any point about it. With your solution in another comment it seems a better solution for it.

On the other hand, I was thinking that `requests` do live in memory, and not coming as stream. But I do definitely prefer to keep the abstraction!